app/lib/service/email/backend.dart

```diff
@@ -0,0 +1,159 @@
+
+/// Represents the backend for the outgoing email queue.
+class EmailBackend {
+  final DatastoreDB _db;
+
+  EmailBackend(this._db);
+
+  /// Creates [OutgoingEmail] entities that can be stored alongside a transaction.
+  List<OutgoingEmail> prepareEntities(EmailMessage msg) {
+    return msg.recipients
+        .map(
+          (recipient) => OutgoingEmail.init(
+            fromEmail: msg.from.email,
+            recipientEmail: recipient.email,
+            subject: msg.subject,
+            bodyText: msg.bodyText,
+          ),
+        )
+        .toList();
+  }
+
+  /// Queries all [OutgoingEmail] objects and tries to send out the email,
+  /// deleting the entry after the email was sent successfully. This method
+  /// should be called only from the background task.
+  ///
+  /// The processing of the Datastore query will be stopped after [stopAfter]
+  /// duration has elapsed. This allows the periodic task to complete within
+  /// the planned time window.
+  ///
+  /// When [ignorePendingAt] is true, the newly created entries are
+  /// sent immediately. This is intended for testing only to force-send
+  /// queued entries.
+  ///
+  /// Returns the number of successfully sent emails.
+  Future<int> trySendAllOutgoingEmails({
+    Duration? stopAfter,
+    @visibleForTesting bool ignorePendingAt = false,
+  }) async {
+    final sw = Stopwatch()..start();
+    final query = _db.query<OutgoingEmail>()..order('-created');
+    var successful = 0;
+    await for (final m in query.run()) {
+      if (stopAfter != null && sw.elapsed > stopAfter) break;
+      if (m.isDead) continue;
+      if (!ignorePendingAt && !m.mayAttemptNow) continue;
+      final wasSent =
+          await _trySendOutgoingEmail(m.uuid, ignorePendingAt: ignorePendingAt);
+      if (wasSent) {
+        successful++;
+      }
+    }
+    return successful;
+  }
+
+  /// Tries to send emails with the given [ids]. The
+  /// [OutgoingEmail] entry will be deleted after the
+  /// email was sent successfully.
+  ///
+  /// This method should be called right after the entries
+  /// are saved in the Datastore.
+  Future<void> trySendOutgoingEmails(Iterable<String> ids) async {
+    for (final id in ids) {
+      await _trySendOutgoingEmail(id, ignorePendingAt: true);
+    }
+  }
+
+  /// Tries to send email with the given [id]. The
+  /// [OutgoingEmail] entry will be deleted after the
+  /// email was sent successfully.
+  ///
+  /// Returns true if the email was sent successfully.
+  Future<bool> _trySendOutgoingEmail(
+    String id, {
+    bool ignorePendingAt = false,
+  }) async {
+    return await withRetryTransaction(_db, (tx) async {
+      final o = await tx.lookupOrNull<OutgoingEmail>(
+          _db.emptyKey.append(OutgoingEmail, id: id));
+      if (o == null || o.isDead) {
+        return false;
+      }
+      if (!ignorePendingAt && !o.mayAttemptNow) {
+        return false;
+      }
+      try {
+        await emailSender.sendMessage(EmailMessage(
```

Comment:
did you notice that you're sending emails inside a `withRetryTransaction` wrapper? That could have some negative implications.

Reply:
Good point, I've separated the DB and email parts, PTAL.